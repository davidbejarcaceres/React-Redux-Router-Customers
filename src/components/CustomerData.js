import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import CustomersActions from './CustomersActions'
import { Button } from '@material-ui/core'
import { accessControl } from '../helpers/accessControl';
import { CUSTOMER_VIEW } from '../constants/permissions';

const CustomerData = ({ id, name, dni, age, onBack, isDeleteAllowed, onDelete }) => {
    return (
        <div>
            <div className="customer-data">
                <h2> Datos del cliente</h2>
                <div><strong>Name: </strong> <i>{name}</i></div>
                <div><strong>DNI: </strong> <i>{dni}</i></div>
                <div><strong>Age</strong> <i>{age}</i></div>
            </div>

            <CustomersActions>
                <Button type="submit" onClick={onBack} variant="contained" color="secundary">
                    Volver
                </Button>

                {isDeleteAllowed &&
                    <Button type="submit" onClick={() => onDelete(id)} variant="contained" color="secundary">
                        Delete
                    </Button>
                }
            </CustomersActions>
        </div>
    )
}

CustomerData.propTypes = {
    name: PropTypes.string.isRequired,
    dni: PropTypes.string.isRequired,
    age: PropTypes.number,
    onBack: PropTypes.func.isRequired,
    isDeleteAllowed: PropTypes.bool,
    onDelete: PropTypes.func
}

export default accessControl([CUSTOMER_VIEW])(CustomerData)
