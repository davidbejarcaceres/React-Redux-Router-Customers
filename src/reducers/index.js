import { combineReducers } from 'redux';
import { customers } from './customers';
import { reducer as reduxForm } from 'redux-form';
import { CUSTOMER_VIEW, CUSTOMER_LIST, CUSTOMER_EDIT } from './../constants/permissions';


export default combineReducers({
    customers,
});