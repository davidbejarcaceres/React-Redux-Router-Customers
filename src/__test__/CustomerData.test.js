import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import CustomerData from '../components/CustomerData';
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import reducers from "../reducers/index"
import { Route, withRouter } from 'react-router-dom';
import { createMemoryHistory } from 'history'

afterEach(cleanup);

const customer = {
    "id": "30000001",
    "dni": "30000001",
    "name": "FACUNDO 2",
    "age": 30
};
const history = createMemoryHistory()

const handleSubmit = async (values) => {
    console.log(JSON.stringify(values));
    // return await props.updateCustomer(values)
}

const handleOnBack = () => {
    history.goBack()
}

const handleSubmitSuccess = () => {
    history.goBack()
}

const handleOnDelete = async (customerID) => {
    console.log("Handle on delete user:" + customerID);     
    // await props.deleteCustomer(customerID)
    // props.history.goBack()
}
const isDelete = true;


// this is a handy function that I normally make available for all my tests
// that deal with connected components.
// you can provide initialState for the entire store that the ui is rendered with
let initialStateStore = {}
function renderWithRedux(
    ui,
    { initialState, store = createStore(reducers, initialStateStore ) } = {}
  ) {
    return {
      ...render(<Provider store={store}>{ui}</Provider>),
      // adding `store` to the returned utilities to allow us
      // to reference it in our tests (just try to avoid using
      // this to test implementation details).
      store,
    }
  }


  test('can render with redux with defaults', () => {
    const { getByText } = renderWithRedux(
        <CustomerData
            id={customer.id}
            name={customer.dni}
            dni={customer.name}
            age={customer.age}
            onSubmit={handleSubmit}
            onSubmitSuccess={handleSubmitSuccess}
            onBack={handleOnBack}
            isDeleteAllowed={!!isDelete}
            onDelete={handleOnDelete}/>
    )
    const customerName = getByText(/FACUNDO 2/i);
    expect(customerName).toBeInTheDocument();
  });


  it("renders Snapshot", () => {
    const { asFragment } = renderWithRedux(
        <CustomerData
            id={customer.id}
            name={customer.dni}
            dni={customer.name}
            age={customer.age}
            onSubmit={handleSubmit}
            onSubmitSuccess={handleSubmitSuccess}
            onBack={handleOnBack}
            isDeleteAllowed={!!isDelete}
            onDelete={handleOnDelete}/>
    );
    expect(asFragment()).toMatchSnapshot();
  });



  
  it("Shows Text from props", () => {
    const { getByText } = renderWithRedux(
        <CustomerData
            id={customer.id}
            name={customer.dni}
            dni={customer.name}
            age={customer.age}
            onSubmit={handleSubmit}
            onSubmitSuccess={handleSubmitSuccess}
            onBack={handleOnBack}
            isDeleteAllowed={!!isDelete}
            onDelete={handleOnDelete}/>
    );
    
    const customerName = getByText(/FACUNDO 2/i);
    const customerAge = getByText("30");
    const customerDNI = getByText(/30000001/i);
    expect(customerName).toBeInTheDocument();
    expect(customerAge).toBeInTheDocument();
    expect(customerDNI).toBeInTheDocument();
  });


  // it("Reacts to clicks functions", () => {
  //   const mockFn = jest.fn();
  //   const handleSubmitTest = () => {
  //       mockFn();
  //   }
  //   const { queryByTestId } = renderWithRedux(
  //       <CustomerData
  //           id={customer.id}
  //           name={customer.dni}
  //           dni={customer.name}
  //           age={customer.age}
  //           onSubmit={handleSubmitTest}
  //           onSubmitSuccess={handleSubmitSuccess}
  //           onBack={handleOnBack}
  //           isDeleteAllowed={!!isDelete}
  //           onDelete={handleOnDelete}/>
  //   );

  //   fireEvent.click(queryByTestId("btn-submit"))
  //   expect(mockFn).toHaveBeenCalled();
  // });

