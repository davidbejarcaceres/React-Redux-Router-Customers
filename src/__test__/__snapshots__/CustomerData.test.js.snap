// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="customer-data"
    >
      <h2>
         Datos del cliente
      </h2>
      <div>
        <strong>
          Name: 
        </strong>
         
        <i>
          30000001
        </i>
      </div>
      <div>
        <strong>
          DNI: 
        </strong>
         
        <i>
          FACUNDO 2
        </i>
      </div>
      <div>
        <strong>
          Age
        </strong>
         
        <i>
          30
        </i>
      </div>
    </div>
    <div>
      <div
        class="customers-actions"
      >
        <div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained"
            data-testid="btn-submit"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Volver
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Delete
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
